module/zfs/dmu_object.c

```diff
@@ -236,34 +236,39 @@ int
 dmu_object_next(objset_t *os, uint64_t *objectp, boolean_t hole, uint64_t txg)
 {
 	uint64_t offset;
-	dmu_object_info_t doi;
+	uint64_t start_obj;
 	struct dsl_dataset *ds = os->os_dsl_dataset;
-	int dnodesize;
 	int error;
 
-	/*
-	 * Avoid expensive dnode hold if this dataset doesn't use large dnodes.
-	 */
-	if (ds && ds->ds_feature_inuse[SPA_FEATURE_LARGE_DNODE]) {
-		error = dmu_object_info(os, *objectp, &doi);
+	if (*objectp == 0) {
+		start_obj = 1;
+	} else if (ds && ds->ds_feature_inuse[SPA_FEATURE_LARGE_DNODE]) {
 		/*
-		 * dmu_object_next() consumers expect ESRCH if the object is not
-		 * allocated.
+		 * For large_dnode datasets, scan from the beginning of the
+		 * dnode block to find the starting offset. This is needed
+		 * because objectp could be part of a large dnode so we can't
+		 * assume it's a hole even if dmu_object_info() returns ENOENT.
 		 */
-		if (error == ENOENT)
-			return (SET_ERROR(ESRCH));
-		else if (error && !(error == EINVAL && *objectp == 0))
-			return (SET_ERROR(error));
-		else
-			dnodesize = doi.doi_dnodesize;
+		int epb = DNODE_BLOCK_SIZE >> DNODE_SHIFT;
+		int skip;
+		uint64_t i;
+
+		for (i = *objectp & ~(epb - 1); i <= *objectp; i += skip) {
+			dmu_object_info_t doi;
+
+			error = dmu_object_info(os, i, &doi);
```

Comment:
To be clear, the early return which was removed can be safely dropped because the code now scans from the beginning on the block which guarantees the passed objectp can never point to the middle of a large dnode.  This allows us to rely on the long standing logic in `dnode_next_offset()` for error handling.